README.md

```diff
@@ -1,78 +1,406 @@
+Generation started at: 08:02:52
+
+Number of rows | Velocity (rows/sec) | Velocity trend
+---------------+---------------------+---------------
+100            | 0                   | - 
+~~~
+
+<!-- bug velocity not right -->
+
+The generator is a command line tool which reads a profile, and outputs data in CSV format. The `--max-rows=100` option tells the generator to create 100 rows of data, and the `--overwrite` option tells it to overwrite previously generated files. The generator outputs progress, in rows per second, which is useful when generating large volumes of data.
+
+If you open up `output.csv` you'll see the following:
+
+~~~
+firstName
+15-09-1971 04:30:18
+28-10-2087 09:23:30
+1876153676
+334227637
+
+03-04-2053 04:57:06
+"Lorem Ipsum"
+[...]
+~~~
+
+The generator has successfully created 100 rows of random data. However, in most cases this data will not be terribly useful. It is likely that `firstName` will only allow string values, with this constraint enforced via a database schema for example. If you don't provide any constraints, the generator will output values from the 'universal set', which  contains all generatable values (null, any string, any date, any number, etc).
+
+Let's assume you only want to generate string values for the `firstName` field; this can be achieved by adding an `ofType` constraint for the field. With this constraint alone, the generator will output random strings containing characters from the unicode [Basic Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane). If you want a more restrictive character set, ths can be achieved by adding `matchingRegex`. 
+
+
+## Adding constraints
+
+Update the JSON profile as follows:
+
+~~~json
+{
+  "schemaVersion": "0.1",
+  "fields": [
+    { "name": "firstName" }
+  ],
+  "rules": [
+    {
+      "rule": "first name constraints",
+      "constraints": [
+        { "field": "firstName", "is": "ofType", "value": "string" },
+        { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" }
+      ]
+    }
+  ]
+}
+~~~
+
+Re-running generation now creates a file containing random strings that match the simple regex `[a-z]{1,10}`:
+
+~~~
+firstName
+"f"
+
+"draxonsa"
+"j"
+"srsub"
+
+"f"
+
+[...]
+~~~
+
+Fields are nullable by default, however, you can add a further constraint to a field to indicate it is 'not null'. If you add the following constraint, the generator will no longer output null values:
+
+~~~json
+"constraints": [
+  { "field": "firstName", "is": "ofType", "value": "string" },
+  { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" },
+  { "not": { "field": "firstName", "is": "null" } }
+]
+~~~
+
+The generator supports three different types of constraint. These are:
+
+ - **Predicates** - boolean-valued functions that define whether a given value as valid or invalid. In the above profile `ofType` and `matchingRegex` are examples of predicates.
```

Comment:
yes ... I am proposing a name change here. I'm not keen on the term 'epistemic'!